--- Marlin-1.1.5/Marlin/pins_TOTUMDUINO_V1.h	1970-01-01 01:00:00.000000000 +0100
+++ Marlin-1.1.5.patched/Marlin/pins_TOTUMDUINO_V1.h	2017-09-01 19:08:54.730765860 +0200
@@ -0,0 +1,141 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (C) 2016 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/**
+ * FABtotum/Totumduino pin assignments
+ */
+
+#if !defined(__AVR_ATmega1280__)
+  #error "Oops!  Make sure you have 'Arduino Mega' selected from the 'Tools -> Boards' menu."
+#endif
+
+#define DEFAULT_MACHINE_NAME    "FABtotum"
+#define DEFAULT_SOURCE_CODE_URL "https://github.com/FABtotum/FABlin-ng"
+#define BOARD_NAME              "Totumduino"
+
+#define X_STEP_PIN         54
+#define X_DIR_PIN          55
+#define X_ENABLE_PIN       38
+#define X_MIN_PIN           2//3
+#define X_MAX_PIN           3//2
+
+#define Y_STEP_PIN         60
+#define Y_DIR_PIN          61
+#define Y_ENABLE_PIN       56
+#define Y_MIN_PIN          14
+#define Y_MAX_PIN          15
+
+#define Z_STEP_PIN         46
+#define Z_DIR_PIN          48
+#define Z_ENABLE_PIN       62
+#define Z_MIN_PIN          19//18
+#define Z_MAX_PIN          18//19
+
+#define Y2_STEP_PIN        36
+#define Y2_DIR_PIN         34
+#define Y2_ENABLE_PIN      30
+
+#define Z2_STEP_PIN        36
+#define Z2_DIR_PIN         34
+#define Z2_ENABLE_PIN      30
+
+// 1st drive: onboard feeder
+#define E0_STEP_PIN        26
+#define E0_DIR_PIN         28
+#define E0_ENABLE_PIN      24
+
+// 2nd drive: 5th axis
+#define E1_STEP_PIN        36
+#define E1_DIR_PIN         34
+#define E1_ENABLE_PIN      30
+
+// 3rd drive (logical): head
+#define E2_STEP_PIN        21
+#define E2_DIR_PIN         20
+#define E2_ENABLE_PIN      23
+
+// 4th drive (logical): bed
+#define E3_STEP_PIN        53
+#define E3_DIR_PIN         34
+#define E3_ENABLE_PIN      8//68
+//#define E3_ENABLE_PIN      68
+
+#define SCAN_STEP_PIN      E3_STEP_PIN
+#define SCAN_ENABLE_PIN    E3_ENABLE_PIN
+#define SCAN_BED_ON_PIN    68
+
+#define LED_PIN            13
+
+#define BEEPER_PIN         4
+#define BEEPER_PIN_INV     1
+
+#define HEATER_0_PIN       10
+#define HEATER_1_PIN       -1
+#define HEATER_2_PIN       -1
+#define HEATER_BED_PIN      8
+#define FAN_PIN             9
+
+#define TEMP_0_PIN         13   // ANALOG NUMBERING
+#define TEMP_BED_PIN       14   // ANALOG NUMBERING
+#define TEMP_1_PIN         -1   // ANALOG NUMBERING
+#define TEMP_2_PIN         -1   // ANALOG NUMBERING
+
+
+//fastio define
+#define WIRE_END_PIN            78
+#define NOT_RASPI_PWR_ON_PIN    85
+#define	NOT_SECURE_SW_PIN       71
+#define	NOT_REEL_LENS_OPEN_PIN  70
+#define	LIGHT_SIGN_ON_PIN       83
+#define	RPI_RECOVERY_PIN        79
+
+//power_off_pins
+#define	PWR_IN_PIN    50
+#define PWR_OUT_PIN   51
+
+
+//additional IO
+#define	RED_PIN             5
+#define	GREEN_PIN           7
+#define	BLUE_PIN            12
+#define	HOT_LED_PIN         53
+#define STAT_LED_RED_PIN    53
+#define	DOOR_OPEN_PIN       49
+#define	HEAD_LIGHT_PIN      45
+#define	LASER_GATE_PIN      44
+#define	I2C_SCL             21
+#define	I2C_SDA             20
+#define	MILL_MOTOR_ON_PIN   41
+#define NOT_SERVO1_ON_PIN   23
+#define	NOT_SERVO2_ON_PIN   25
+
+//additional analog
+#define MAIN_CURRENT_SENSE_PIN  12  // ANALOG NUMBERING
+#define	MON_5V_PIN              10  // ANALOG NUMBERING
+#define	MON_24V_PIN             9   // ANALOG NUMBERING
+#define	PRESSURE_ANALOG_PIN     3   // ANALOG NUMBERING
+
+//servo
+#define SERVO0_PIN         11
+#define SERVO1_PIN         6
+
+// Default pins for additional serial ports (may be reconfigured at run-time)
+#define RX4 67
+#define TX4 11
