--- Marlin-1.1.5/Marlin/boards.h	2017-08-29 01:46:39.000000000 +0200
+++ Marlin-1.1.5.patched/Marlin/boards.h	2017-09-01 02:59:06.887319406 +0200
@@ -100,6 +100,7 @@
 #define BOARD_BAM_DICE_DUE      402  // 2PrintBeta BAM&DICE Due with STK drivers
 #define BOARD_BQ_ZUM_MEGA_3D    503  // bq ZUM Mega 3D
 #define BOARD_ZRIB_V20          504  // zrib V2.0 control board (Chinese knock off RAMPS replica)
+#define BOARD_TOTUMDUINO_V1     25   // FABtotum totumduino board v1.0
 
 #define MB(board) (MOTHERBOARD==BOARD_##board)
 
